### PR TITLE
Use filtered informer to watch OIDC service accounts

### LINKED ISSUE #3655: [control-plane]: Use filtered informer to watch OIDC service accounts
Currently we're watching all service accounts in the cluster for changes and reenque the objects which have an OIDC service account assigned if something changes:

https://github.com/knative-extensions/eventing-kafka-broker/blob/45c65c8b780a312de497e560e5afccb2644feca9/control-plane/pkg/reconciler/trigger/controller.go#L145-L149

and 

https://github.com/knative-extensions/eventing-kafka-broker/blob/45c65c8b780a312de497e560e5afccb2644feca9/control-plane/pkg/reconciler/trigger/namespaced_controller.go#L141-L144

Instead we should use a filtered service account informer as done in https://github.com/knative/eventing/issues/7341

**Additional information:**
* requires #3521 done first
* requires #3658 
* In https://github.com/knative/eventing/pull/7527 we did this in eventing-core

### PR DESCRIPTION
**Pull request by karkabbage**
Fixes #3655 

## Proposed Changes

- Intial changes similar to #[7527](https://github.com/knative/eventing/pull/7527/files) to manage filtered accounts
-
-

**Release Note**

```release-note

```

**Docs**

### CODE REVIEW

**Comment by creydr** [comment]
Hey @karkabbage,
Thanks for your work so far. Can you rebase this PR to solve the merge conflicts? Also it seems you're missing to run `./hack/update-codegen.sh` or `./hack/update-deps.sh` to vendor the newly imported packages.

**Comment by karkabbage** [comment]
Hi @creydr , 
I've rebased the merge conflicts and ran codegen, but no new files were made. 
Is this expected ? I attempted using 	- "knative.dev/pkg/client/injection/kube/informers/factory/filtered/fake"
in controller_test.go
Thanks!

**Comment by karkabbage** [comment]
@knative-prow-robot /retest

File: control-plane/pkg/reconciler/trigger/controller_test.go [hunk 1]
```diff
@@ -56,8 +63,14 @@ func TestNewController(t *testing.T) {
...
 
+func SetUpInformerSelector(ctx context.Context) context.Context {
+	ctx = filteredFactory.WithSelectors(ctx, auth.OIDCLabelSelector)
+	return ctx
+}
+
 func TestFilterTriggers(t *testing.T) {
-	ctx, _ := reconcilertesting.SetupFakeContext(t)
+	// ctx, _ := reconcilertesting.SetupFakeContext(t)
+	ctx, _ := SetupFakeContext(t, SetUpInformerSelector)
```
Comment (creydr) [inline control-plane/pkg/reconciler/trigger/controller_test.go:73]:
```suggestion
	ctx, _ := reconcilertesting.SetupFakeContext(t, SetUpInformerSelector)
```

File: control-plane/pkg/reconciler/trigger/namespaced_controller_test.go [hunk 2]
```diff
@@ -37,7 +41,8 @@ import (
 )
 
 func TestNewNamespacedController(t *testing.T) {
-	ctx, _ := reconcilertesting.SetupFakeContext(t)
+	// ctx, _ := reconcilertesting.SetupFakeContext(t)
+	ctx, _ := SetupFakeContext(t, SetUpInformerSelector)
```
Comment (creydr) [inline control-plane/pkg/reconciler/trigger/namespaced_controller_test.go:45]:
```suggestion
	ctx, _ := reconcilertesting.SetupFakeContext(t, SetUpInformerSelector)
```

**Review by creydr — CHANGES_REQUESTED** [review]
To get rid of all the Go-style check, you can run `./hack/update-codegen.sh` (and push the changes)

File: control-plane/pkg/reconciler/trigger/controller.go [hunk 3]
```diff
@@ -153,7 +156,8 @@ func NewController(ctx context.Context, watcher configmap.Watcher, configs *conf
 	secretinformer.Get(ctx).Informer().AddEventHandler(controller.HandleAll(reconciler.Tracker.OnChanged))
 
 	// Reconciler Trigger when the OIDC service account changes
-	serviceaccountInformer.Informer().AddEventHandler(cache.FilteringResourceEventHandler{
+	// serviceaccountInformer.Informer().AddEventHandler(cache.FilteringResourceEventHandler{
+		oidcServiceaccountInformer.Informer().AddEventHandler(cache.FilteringResourceEventHandler{
 		FilterFunc: controller.FilterController(&eventing.Trigger{}),
 		Handler:    controller.HandleAll(impl.EnqueueControllerOf),
```
Comment (pierDipi) [inline control-plane/pkg/reconciler/trigger/controller.go:162]: @creydr we need to make sure that the trigger belogs to a broker with "our" broker class

File: control-plane/pkg/reconciler/trigger/namespaced_controller.go [hunk 4]
```diff
@@ -150,7 +155,8 @@ func NewNamespacedController(ctx context.Context, watcher configmap.Watcher, con
 	secretinformer.Get(ctx).Informer().AddEventHandler(controller.HandleAll(reconciler.Tracker.OnChanged))
 
 	// Reconciler Trigger when the OIDC service account changes
-	serviceaccountInformer.Informer().AddEventHandler(cache.FilteringResourceEventHandler{
+	// serviceaccountInformer.Informer().AddEventHandler(cache.FilteringResourceEventHandler{
+	oidcServiceaccountInformer.Informer().AddEventHandler(cache.FilteringResourceEventHandler{
 		FilterFunc: controller.FilterController(&eventing.Trigger{}),
 		Handler:    controller.HandleAll(impl.EnqueueControllerOf),
```
Comment (pierDipi) [inline control-plane/pkg/reconciler/trigger/namespaced_controller.go:161]: ditto here

**Comment by creydr** [comment]
Hi @karkabbage and @yijie-04,
we had a discussion in https://cloud-native.slack.com/archives/C04LMU33V1S/p1712135067227149 about the test failures and @pierDipi found, that it is due to the import of `knative.dev/eventing-kafka-broker/control-plane/pkg/reconciler/trigger` in [trigger_finalizer_test.go](https://github.com/knative-extensions/eventing-kafka-broker/blob/2e15ce667b08417bd9f8784bb8b4480bbe938a3e/test/e2e_new/trigger_finalizer_test.go#L40).
When you remove this import and replace its usage https://github.com/knative-extensions/eventing-kafka-broker/blob/2e15ce667b08417bd9f8784bb8b4480bbe938a3e/test/e2e_new/trigger_finalizer_test.go#L104
with the constants value (`kafka.triggers.eventing.knative.dev`)
```
require.NotEqual(t, f, "kafka.triggers.eventing.knative.dev", "%+v", tr)
```
you should be fine.

**Comment by creydr** [comment]
hello @karkabbage @yijie-04,
any updates on this? Do you need more help?

**Comment by karkabbage** [comment]
> hello @karkabbage @yijie-04, any updates on this? Do you need more help?

Hi I'll be making the changes today!

File: control-plane/pkg/reconciler/trigger/controller.go (hunk 3 above)
Comment (creydr) [reply control-plane/pkg/reconciler/trigger/controller.go:162]: @karkabbage you can check on https://github.com/knative/eventing/pull/7849, how this was done in eventing-core

**Comment by pierDipi** [comment]
@creydr what's the status here? can you take a look?

**Comment by creydr** [comment]
@karkabbage @yijie-04 do you need any more information or help besides https://github.com/knative-extensions/eventing-kafka-broker/pull/3719#discussion_r1580549978?

File: control-plane/pkg/reconciler/trigger/controller.go (hunk 3 above)
Comment (creydr) [reply control-plane/pkg/reconciler/trigger/controller.go]: Addressed in 3e9129d0e60dd92de599837b8b80ee66bba772e6

File: control-plane/pkg/reconciler/trigger/namespaced_controller.go (hunk 4 above)
Comment (creydr) [reply control-plane/pkg/reconciler/trigger/namespaced_controller.go]: Addressed in 971f43b8052a2aca24ab2b74fe15f4d166582143

**Comment by creydr** [comment]
@pierDipi @Cali0707 @Leo6Leo as I commited to this PR: could you PTAL?

File: control-plane/pkg/reconciler/trigger/namespaced_controller.go [hunk 5]
```diff
@@ -150,8 +152,8 @@ func NewNamespacedController(ctx context.Context, watcher configmap.Watcher, con
 	secretinformer.Get(ctx).Informer().AddEventHandler(controller.HandleAll(reconciler.Tracker.OnChanged))
 
 	// Reconciler Trigger when the OIDC service account changes
-	serviceaccountInformer.Informer().AddEventHandler(cache.FilteringResourceEventHandler{
-		FilterFunc: controller.FilterController(&eventing.Trigger{}),
+	oidcServiceaccountInformer.Informer().AddEventHandler(cache.FilteringResourceEventHandler{
+		FilterFunc: filterOIDCServiceAccounts(triggerInformer.Lister(), brokerInformer.Lister(), kafka.BrokerClass, FinalizerName),
```
Comment (Cali0707) [inline control-plane/pkg/reconciler/trigger/namespaced_controller.go:156]: Should this be the namespaced broker class here instead of the normal one?
```suggestion
		FilterFunc: filterOIDCServiceAccounts(triggerInformer.Lister(), brokerInformer.Lister(), kafka.NamespacedBrokerClass, FinalizerName),
```

File: test/e2e_new/trigger_finalizer_test.go [hunk 6]
```diff
@@ -101,7 +99,7 @@ func hasNoKafkaBrokerFinalizer() func(ctx context.Context, t feature.T) {
 		time.Sleep(time.Second * 20) // "eventually"
 		tr := triggerfeatures.GetTrigger(ctx, t)
 		for _, f := range tr.Finalizers {
-			require.NotEqual(t, f, triggerreconciler.FinalizerName, "%+v", tr)
+			require.NotEqual(t, f, "kafka.triggers.eventing.knative.dev", "%+v", tr)
```
Comment (Cali0707) [inline test/e2e_new/trigger_finalizer_test.go:102]: what's the need for this change?

Comment (creydr) [reply test/e2e_new/trigger_finalizer_test.go:102]: Full context is here: https://github.com/knative-extensions/eventing-kafka-broker/pull/3719#issuecomment-2034067621.
TL;DR; is, that this lead to an import an then registered an informer, which lead to a test failure in this case.

**Review by Cali0707 — APPROVED** [review]
/lgtm

**Comment by creydr** [comment]
`wait.go:200: test-mrqmxoln/containersource-eoaibzgi condition is {"type":"Ready","status":"False","lastTransitionTime":"2024-05-06T15:08:59Z","reason":"TrustBundlePropagation","message":"failed to create ConfigMap test-mrqmxoln/knative-eventing-bundlekne-bundle: configmaps \"knative-eventing-bundlekne-bundle\" already exists"}`

/retest-required
